homeassistant/components/hue_ble/config_flow.py

```diff
@@ -0,0 +1,186 @@
+"""Config flow for Hue BLE integration."""
+
+from __future__ import annotations
+
+import logging
+from typing import Any
+
+from HueBLE import HueBleLight
+import voluptuous as vol
+
+from homeassistant.components import bluetooth
+from homeassistant.components.bluetooth.api import (
+    async_ble_device_from_address,
+    async_scanner_count,
+)
+from homeassistant.config_entries import ConfigFlow, ConfigFlowResult
+from homeassistant.const import CONF_MAC, CONF_NAME
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import HomeAssistantError
+from homeassistant.helpers import device_registry as dr
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+STEP_USER_DATA_SCHEMA = vol.Schema(
+    {vol.Required(CONF_NAME): str, vol.Required(CONF_MAC): str}
+)
+
+
+async def validate_input(hass: HomeAssistant, data: dict[str, Any]) -> None:
+    """Validate the user input allows us to connect."""
+
+    address = data[CONF_MAC]
+    ble_device = async_ble_device_from_address(hass, address.upper(), connectable=True)
+
+    if ble_device is None:
+        count_scanners = async_scanner_count(hass, connectable=True)
+        _LOGGER.debug("Count of BLE scanners in HA bt: %i", count_scanners)
+
+        if count_scanners < 1:
+            raise ScannerNotAvailable
+        raise NotFound
+
+    try:
+        light = HueBleLight(ble_device)
+
+        await light.connect()
+        if not light.authenticated:
+            raise InvalidAuth
+
+        if not light.connected:
+            raise CannotConnect
+
+        state_changed, errors = await light.poll_state()
+        if not len(errors) == 0:
+            raise CannotConnect
+    finally:
+        await light.disconnect()
+
+
+class HueBleConfigFlow(ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for Hue BLE."""
+
+    VERSION = 1
+
+    def __init__(self) -> None:
+        """Initialize the config flow."""
+        self._discovery_info: bluetooth.BluetoothServiceInfoBleak | None = None
+
+    async def async_step_user(
+        self,
+        user_input: dict[str, Any] | None = None,
+    ) -> ConfigFlowResult:
+        """Handle manually setting up a light."""
+
+        errors: dict[str, str] = {}
+        if user_input is None:
+            return self.async_show_form(
+                step_id="user", data_schema=STEP_USER_DATA_SCHEMA, errors=errors
+            )
+
+        try:
```

Comment:
I implemented this in an earlier revision but I wasn't able to get it to work reliably, it would produce the list fine but when shown in the UI the entries in the list would be present and select-able but would be blank, i.e no text was shown where the name and mac should have been and I was not able to find a fix and was running out of time I could justify spending on it, especially since it is not the intended way to use the integration, more as a backup in case automatic Bluetooth discovery was not working for some reason.

Reply:
If you don't cache the discovered devices in `self`, they won't carry over to the next step. This example user flow should work with some small adjustments:

```
    async def async_step_user(
        self, user_input: dict[str, Any] | None = None
    ) -> ConfigFlowResult:
        """Handle the user step to pick discovered device."""
        if user_input is not None:
            address = user_input[CONF_ADDRESS]
            title = self._discovered_devices[address]
            await self.async_set_unique_id(address, raise_on_progress=False)
            self._abort_if_unique_id_configured()
            return self.async_create_entry(title=title, data={})

        current_addresses = self._async_current_ids()
        for discovery_info in async_discovered_service_info(self.hass, True):
            address = discovery_info.address
            if (
                DISCOVERY_SVC_UUID not in discovery_info.service_uuids
                or address in current_addresses
                or address in self._discovered_devices
            ):
                continue
            self._discovered_devices[address] = discovery_info.name

        if not self._discovered_devices:
            return self.async_abort(reason="no_devices_found")

        return self.async_show_form(
            step_id="user",
            data_schema=self.add_suggested_values_to_schema(vol.Schema(
                {vol.Required(CONF_ADDRESS): vol.In(self._discovered_devices)}
            ), user_input
            ),
        )
```

Reply:
I implemented this in a test version but it highlighted a problem I encountered previously but forgotten, using this manual method over the Bluetooth discovery method does not work if the Bluetooth discovery step detects the light since that flow will already be in progress, meaning the manual flow with the name and MAC fields will not be able to progress, which makes automatic filling of the form kind of pointless, I guess I could modify it so Bluetooth discovery does not set the ID and use some other mechanism to avoid duplicates but that seems overkill, maybe just edit the strings to point out the manual form with the name and MAC should only be used if the automatic Bluetooth discovery doesn't detect it?

Here is the code I added to test this:
``` 
       errors: dict[str, str] = {}
        if user_input is None:
            current_addresses = self._async_current_ids()
            for discovery_info in async_discovered_service_info(self.hass, True):
                address = discovery_info.address
                if (
                    UUID_HUE_IDENTIFIER in discovery_info.service_uuids
                    and dr.format_mac(address) not in current_addresses
                ):  # or address in self._discovered_devices
                    return self.async_show_form(
                        step_id="user",
                        data_schema=self.add_suggested_values_to_schema(
                            STEP_USER_DATA_SCHEMA,
                            {CONF_NAME: discovery_info.name, CONF_MAC: address},
                        ),
                        description_placeholders={"url_pairing_mode": URL_PAIRING_MODE},
                        errors=errors,
                    )
```